cmd/api/auth/authn.go

```diff
@@ -0,0 +1,54 @@
+// Copyright KubeArchive Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package auth
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	authnv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientAuthnv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+)
+
+func extractBearerToken(header string) (string, error) {
+	if header == "" {
+		return "", errors.New("bad authorization bearer token given")
+	}
+
+	jwtToken := strings.Split(header, " ")
+	if len(jwtToken) != 2 {
+		return "", errors.New("incorrectly formatted authorization header")
+	}
+
+	return jwtToken[1], nil
+}
+
+func Authentication(tri clientAuthnv1.TokenReviewInterface) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		token, err := extractBearerToken(c.GetHeader("Authorization"))
+		if err != nil {
+			abort(c, err.Error(), http.StatusBadRequest)
+			return
+		}
+		tr, err := tri.Create(c, &authnv1.TokenReview{
+			Spec: authnv1.TokenReviewSpec{
+				Token: token,
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			abort(c, fmt.Sprintf("Unexpected error on TR: %s", err.Error()), http.StatusInternalServerError)
+			return
+		}
+		if !tr.Status.Authenticated {
+			abort(c, "Unauthenticated", http.StatusUnauthorized)
```

Comment:
Should this error message be "Unauthenticated" or should it be "Authentication failed"?

Reply:
We should use `http.StatusText(http.StatusUnauthorized)` if we don't want to use a custom text message.

Reply:
I've chosen to use "Authentication failed" which is more accurate to what is happening. I don't want to set it as "Unauthorized" with `http.StatusText(http.StatusUnauthorized)` because the problem here would be the authentication and not the authorization.